libraries/exoplayer_rtsp/src/main/java/androidx/media3/exoplayer/rtsp/reader/RtpOpusReader.java

```diff
@@ -0,0 +1,152 @@
+
+import static androidx.media3.common.util.Assertions.checkArgument;
+import static androidx.media3.common.util.Assertions.checkStateNotNull;
+
+import androidx.media3.common.C;
+import androidx.media3.common.Format;
+import androidx.media3.common.util.Log;
+import androidx.media3.common.util.ParsableByteArray;
+import androidx.media3.common.util.Util;
+import androidx.media3.exoplayer.rtsp.RtpPacket;
+import androidx.media3.exoplayer.rtsp.RtpPayloadFormat;
+import androidx.media3.extractor.ExtractorOutput;
+import androidx.media3.extractor.OpusUtil;
+import androidx.media3.extractor.TrackOutput;
+import java.util.List;
+import org.checkerframework.checker.nullness.qual.MonotonicNonNull;
+
+/**
+ * Parses an OPUS byte stream carried on RTP packets and extracts individual samples. Refer to
+ * RFC7845 for more details.
+ */
+/* package */ final class RtpOpusReader implements RtpPayloadReader {
+  private static final String TAG = "RtpOpusReader";
+
+  private final RtpPayloadFormat payloadFormat;
+  private @MonotonicNonNull TrackOutput trackOutput;
+  private long firstReceivedTimestamp;
+  private long startTimeOffsetUs;
+
+  private final int sampleRate;
+  private int previousSequenceNumber;
+  private boolean foundOpusIDHeader;
+  private boolean foundOpusCommentHeader;
+
+  public RtpOpusReader(RtpPayloadFormat payloadFormat) {
+    this.payloadFormat = payloadFormat;
+    this.firstReceivedTimestamp = C.INDEX_UNSET;
+    this.sampleRate = this.payloadFormat.clockRate;
+    this.previousSequenceNumber = C.INDEX_UNSET;
+    this.foundOpusIDHeader = false;
+    this.foundOpusCommentHeader = false;
+  }
+
+  // RtpPayloadReader implementation.
+
+  @Override
+  public void createTracks(ExtractorOutput extractorOutput, int trackId) {
+    trackOutput = extractorOutput.track(trackId, C.TRACK_TYPE_AUDIO);
+    trackOutput.format(payloadFormat.format);
+  }
+
+  @Override
+  public void onReceivingFirstPacket(long timestamp, int sequenceNumber) {
+    this.firstReceivedTimestamp = timestamp;
+  }
+
+  @Override
+  public void consume(
+      ParsableByteArray data, long timestamp, int sequenceNumber, boolean rtpMarker) {
+    checkStateNotNull(trackOutput);
+
+    /* RFC7845 Section 3
+     * +---------+ +----------------+ +--------------------+ +-----
+     * |ID Header| | Comment Header | |Audio Data Packet 1 | | ...
+     * +---------+ +----------------+ +--------------------+ +-----
+     */
+    if (!foundOpusIDHeader) {
+      int currPosition = data.getPosition();
+      checkArgument(isOpusIDHeader(data), "ID Header missing");
+
+      data.setPosition(currPosition);
+      List<byte[]> initializationData = OpusUtil.buildInitializationData(data.getData());
+      Format.Builder formatBuilder = payloadFormat.format.buildUpon();
+      formatBuilder.setInitializationData(initializationData);
+      trackOutput.format(formatBuilder.build());
+      foundOpusIDHeader = true;
+    } else if (!foundOpusCommentHeader) {
```

Comment:
What about checking ID header and comment header in one go? given they should be together in the packet.

Reply:
We haven't been able to generate a streaming data with both headers being a part of same packet. Can you suggest something for validating the corresponding code changes?